diff --git a/floating-ui-react/middleware/arrow.js b/floating-ui-react/middleware/arrow.js
index e228afafb7a0a12f70c73e978418eadbe334091d..bf1ef0b9df3656d1f9cd10aa5077badb0b8d7f89 100644
--- a/floating-ui-react/middleware/arrow.js
+++ b/floating-ui-react/middleware/arrow.js
@@ -95,8 +95,14 @@ const baseArrow = options => ({
  * @see https://floating-ui.com/docs/arrow
  */
 exports.baseArrow = baseArrow;
-const arrow = (options, deps) => ({
-  ...baseArrow(options),
-  options: [options, deps]
-});
+const arrow = (options, deps) => {
+  const middleware = baseArrow(options);
+  return {
+    name: middleware.name,
+    async fn(state) {
+      return middleware.fn(state);
+    },
+    options: [options, deps]
+  };
+};
 exports.arrow = arrow;
diff --git a/floating-ui-react/middleware/arrow.mjs b/floating-ui-react/middleware/arrow.mjs
index 6d72afc331e3c24d734a1617a4653c477b03c169..039e6b10cb015ffc1c98bf8943db23487b67e0f5 100644
--- a/floating-ui-react/middleware/arrow.mjs
+++ b/floating-ui-react/middleware/arrow.mjs
@@ -88,7 +88,13 @@ export const baseArrow = options => ({
  * This wraps the core `arrow` middleware to allow React refs as the element.
  * @see https://floating-ui.com/docs/arrow
  */
-export const arrow = (options, deps) => ({
-  ...baseArrow(options),
-  options: [options, deps]
-});
\ No newline at end of file
+export const arrow = (options, deps) => {
+  const middleware = baseArrow(options);
+  return {
+    name: middleware.name,
+    async fn(state) {
+      return middleware.fn(state);
+    },
+    options: [options, deps]
+  };
+};
